File: learnLeetCode_remote/Day21/Part2/shortest_path_visiting_all_nodes.py
# LC 847 — Shortest Path Visiting All Nodes
# BFS with state (node, mask)
from collections import deque
from typing import List

def shortestPathLength(graph: List[List[int]]) -> int:
    n = len(graph)
    if n <= 1:
        return 0
    allmask = (1 << n) - 1
    q = deque()
    visited = [[False]*(1<<n) for _ in range(n)]
    for i in range(n):
        mask = 1 << i
        q.append((i, mask, 0))
        visited[i][mask] = True
    while q:
        u, mask, d = q.popleft()
        if mask == allmask:
            return d
        for v in graph[u]:
            nmask = mask | (1 << v)
            if not visited[v][nmask]:
                visited[v][nmask] = True
                q.append((v, nmask, d+1))
    return -1

if __name__ == "__main__":
    print(shortestPathLength([[1,2,3],[0],[0],[0]]))  # 4

File: learnLeetCode_remote/Day21/Part2/shortest_path_visiting_all_nodes.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 847 — Shortest Path Visiting All Nodes
// BFS with state (node, mask)

int shortestPathLength(vector<vector<int>>& graph){
    int n = graph.size();
    if (n <= 1) return 0;
    int allmask = (1<<n) - 1;
    queue<tuple<int,int,int>> q; // node, mask, dist
    vector<vector<char>> vis(n, vector<char>(1<<n, 0));
    for(int i=0;i<n;++i){
        int m = 1<<i; vis[i][m]=1; q.emplace(i, m, 0);
    }
    while(!q.empty()){
        auto [u, mask, d] = q.front(); q.pop();
        if (mask == allmask) return d;
        for(int v: graph[u]){
            int nmask = mask | (1<<v);
            if (!vis[v][nmask]){ vis[v][nmask]=1; q.emplace(v, nmask, d+1); }
        }
    }
    return -1;
}

int main(){
    vector<vector<int>> g = {{1,2,3},{0},{0},{0}};
    cout << shortestPathLength(g) << "\n"; // 4
    return 0;
}
